stacker/variables.py

```diff
@@ -140,21 +128,358 @@ def resolve(self, context, provider):
                 the base provider
 
         """
+        try:
+            self._value.resolve(context, provider)
+        except FailedLookup as e:
+            raise FailedVariableLookup(self.name, e.lookup, e.error)
 
-        while self.lookups:
-            resolved_lookups = resolve_lookups(self, context, provider)
-            self.replace(resolved_lookups)
+    def dependencies(self):
+        """
+        :return: list of stack names that this variable depends on
+        :rtype: Set[str]
+        """
+        return self._value.dependencies()
 
-    def replace(self, resolved_lookups):
-        """Replace lookups in the Variable with their resolved values.
 
-        Args:
-            resolved_lookups (dict): dict of :class:`stacker.lookups.Lookup` ->
-                resolved value.
+class VariableValue(object):
+    """
+    Abstract Syntax Tree base object to parse the value for a variable
+    """
+    def value(self):
+        return NotImplementedError()
+
+    def __iter__(self):
+        return NotImplementedError()
+
+    def resolved(self):
+        """
+        :return: Whether value() will not raise an error
```

Comment:
Can you convert these to the Google style docstrings, that's the standard we use in stacker.